### DISCUSSION TITLE
setsockopt function OSError:[Errno 22] EINVAL

### DISCUSSION THREAD

**Discussion by skylin008** [micropython/micropython#13225] [STM32 / Pyboard] [answered]
I had rebuild the firmware from  the Micropython v1.22 version, when used the socket setsockopt function ,shows: 
Task exception wasn't retrieved
future: <Task> coro= <generator object 'test' at 200047f0>
Traceback (most recent call last):
  File "asyncio/core.py", line 1, in run_until_complete
  File "<stdin>", line 198, in test
  File "<stdin>", line 81, in __init__
OSError: [Errno 22] EINVAL

# used w5500 for ethernet commuication chip
# use udp protocol

from machine import SPI, Pin
import gc
import socket
import uasyncio as asyncio
import logging
import network
from primitives.ringbuf_queue import RingbufQueue
from micropython import alloc_emergency_exception_buf
from time import sleep_ms

gc.collect()

from InitConfig import config

log = logging.MiniLog("Ethernet", level = logging.DEBUG)

alloc_emergency_exception_buf(100)

DEBUG = config.get('DEBUG', False)

__version__ = "1.0"
__author__  = "skylin"

class Ethernet():
    ip          = config.get('IP')        # IP address
    subnet      = config.get('SUBNET')    # subnet_mask
    gateway     = config.get('GATEWAY')   # gateway
    dns         = config.get('DNS')       # dns   
    
    def __init__(self,  maxtry = 100):
        _spi = SPI(1, baudrate = 10_000_000)    # SPI1 highest speed for 42M
        _cs  = Pin('E_SS', Pin.OUT)
        _rst = Pin('E_RST', Pin.OUT)
        _rst(0)
        sleep_ms(50)
        _rst(1)
        self._maxtry = maxtry
        self._isRuning = False
        self.nic = network.WIZNET5K(_spi, _cs, _rst)
        self.nic.active(False)
        self.nic.active(True)
        self.nic.ifconfig((self.ip, self.subnet, self.gateway, self.dns))
        cnt = 0
        while not self.nic.isconnected():
            cnt = cnt + 1
            sleep_ms(200)
            if maxtry is not None and cnt >= maxtry:
                cnt = 0
                break
        self._isRuning = True
    
    def status(self):
        return self._isRuning

class Udp():

    def __init__(self):

        host        = config.get('HOST')      # server IP
        port        = config.get('PORT')      # server port

        
        self.ringbuf = RingbufQueue([0 for _ in range(100)])
        
        self.isRuning = False
        self.eth = Ethernet()
        print(self.eth.status())
        if self.eth.status():
            s    = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)  # make a udp socket
            addr = socket.getaddrinfo(host, port)[0][-1]
            s.bind(('',port))                                        # bind porrt
            s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)  # if netlink free can be used for next time
        
            s.setblocking(False)
            self._sock = s
            self.addr  = addr
            self.isRuning = True
            #self.sreader = asyncio.StreamReader(self._sock)
            #self.swriter = asyncio.StreamWriter(self._sock, {})

    def close(self):
        self._sock.close()
    
    def status(self):
        return self.isRuning

    
    def write(self):
        if self.isRuning: 
            if not self.ringbuf.empty():
                try:
                    data = self.ringbuf.get_nowait()
                    print(data) 
                                      
                    self._sock.sendto(data, self.addr)

                except OSError as e:
                    self.close()
                finally:
                    pass

    @micropython.native
    def frame(self, data):
        if data is not None:
            self.ringbuf.put_nowait(data)
            #res = self.ringbuf.get_nowait()
            #print(res)
    

    async def run(self):
        if self.status():
            if not self.ringbuf.empty():

                data = self.ringbuf.get_nowait()
                print(f"state:{self.state()}")
                print(f"sendData:{data}")
                
                try:
                    await self.swriter.awrite(data)
                    await self.swriter.drain()
                except:
                    self.close()
                    return 
             

class Frame(object):

    NET_PACKET_HEADER    = config.get('NET_PACKET_HEADER')
    NET_PACKET_TYPE      = config.get('NET_PACKET_TYPE')
    NET_PACKET_COMPID    = config.get('NET_PACKET_COMPID')
    NET_PACKET_TIMESTAMP = config.get('NET_PACKET_TIMESTAMP')
    NET_PACKET_END       = config.get('NET_PACKET_END')

    def __init__(self):
        pass

    @micropython.native
    def _parity_check(self, bytecode):
        """
        even return True
        odd  return False
        """
        parity = 0
        for b in bytecode:
            parity ^= b & 1
            for i in range(1, 8):
                parity ^= (b >> i) & 1
        return parity == 0

    @micropython.native
    def pack(self, distance, angle):               
       
        d = b''.join(map(lambda x:int.to_bytes(x,2,'little'),distance))

        a = b''.join(map(lambda x:int.to_bytes(x,1,'little'),angle))
        
        dataPacket = d + a
    
        dataLength = len(dataPacket)  
    
        packetLength = (42 + dataLength).to_bytes(8, 'little')   
    
        if (self._parity_check(dataPacket)):
            parity = b'\x01'
        else:
            parity = b'\x00'
            
        msg = self.NET_PACKET_HEADER + packetLength + self.NET_PACKET_TYPE + self.NET_PACKET_COMPID + self.NET_PACKET_TIMESTAMP + dataPacket + parity + self.NET_PACKET_END 
        
        print(f"package:{msg}")
        return msg
    

if __name__ == '__main__':
    
    async def test():
            
        data = [10,20,30,40,50,60]
        ang  = [10,20,30]
        f = Frame()
        msg = f.pack(data, ang)
        u = Udp()
        i = 0
        while True:
            u.frame(msg)
            u.write()
            
            
            await asyncio.sleep_ms(200)
    
    async def start():
        asyncio.create_task(test())
        while True:
            await asyncio.sleep_ms(10)

    app = None
    async def main():
        import time
        t = time.localtime()
        log.debug(t)
        gc.collect()
        global app
        
        await start()

    try:
        gc.collect()
        asyncio.run(main())

    except KeyboardInterrupt:
        log.info('Interrupt')

    finally:
        asyncio.new_event_loop()
        log.info("run again")

`

**Comment by zcattacz** [comment]
addr = socket.getaddrinfo(host, port)[0][-1]
        s.bind(('',port))                                        # bind porrt
        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)  # if netlink free can be used for next time

`bind( ("",port) )` shouldn't work, at least never worked for me.
always have to use the bytes sequence returned by getaddrinfo() in micropython.

**Reply by zcattacz** [reply, marked as answer]
Also `SO_REUSEADDR` needs to be set before bind() to actually work.

**Reply by skylin008** [reply]
Thank you @zcattacz , I had set sock.setsockopt before socket.bind, the same error show.

**Reply by zcattacz** [reply]
did you fix the bind() call? `bind(addr)`, instead of `bind(("",port))`

**Reply by skylin008** [reply]
thanks @zcattacz , I had bind(addr), the same error show when used  setsockopt function.
